### PR TITLE
k2: refactor telnet systest to avoid race condition

### PR DESCRIPTION
**Pull request by Risca**
This change, together with #21, should make this systest a bit more stable. I never had an issue on Arch Linux, but on Ubuntu Server, the external zk2 process managed to exit before the client received the `SUT_RECOVERY_PERFORM` message.

### CODE REVIEW

**Comment by per-bohlin** [comment]
@Risca 
I'm feeling a bit dense. I don't see how the changes would change anything. Based on my understanding of the existing code, your changes doesn't add anything. Please elaborate on how your fix works and why the existing code is broken.

**Comment by Risca** [comment]
> @Risca
> I'm feeling a bit dense. I don't see how the changes would change anything. Based on my understanding of the existing code, your changes doesn't add anything. Please elaborate on how your fix works and why the existing code is broken.

If I've understood it correctly, this is what's happening on my Ubuntu server:
1. Remote (zk2 process) starts up and blocks on init
2. Client (systest) starts up and connects to remote
3. Client creates a local message queue and listens for `SUT_RECOVERY_PERFORM`
4. Client unblocks the remote
5. The remote does a connectivity check, fails and sends a `SUT_RECOVERY_PERFORM`
6. The remote does a second connectivity check, fails and disconnects
7. The client gets an `EOFException` due to the remote being disconnected _before_ if receives the `SUT_RECOVERY_PERFORM` message.
8. The systest fails

This happens occasionally, i.e. this is not 100 % reproducible.

I'm not exactly sure how the RemoteBlocker component works, but this is what I believe this change does:

4. Client unblocks the remote
5. The remote does a connectivity check, fails and then gets blocked on sending `SUT_RECOVERY_PERFORM`
6. The client unblocks the remote, which does a second connectivity check, fails and disconnects.
7. Since the timing has changed, the remote has time to both receive the `SUT_RECOVERY_PERFORM` and `wait_for_finished()` before any `EOFException`.

I don't particularly like this last part. I was thinking of rewriting the systest a bit more and only require that `blocker.wait_for_started()` succeeds:
```python
blocker.wait_for_started(timeout=10)

# continue execution
blocker.stop_blocking()
try:
    # expect SUT_RECOVERY_PERFORM
    queue.get(timeout=10)

    assert blocker.wait_for_finished(timeout=1)
except:
    # log this?
    pass
```
That is if my understanding of `RemoteBlocker` is correct :smile_cat:

**Comment by Risca** [comment]
This is the error output I occasionally get:
```
Failures/Errors:
----------------
telnet.systest.test_telnet.test_sut_recovery_initiated_when_telnet_connection_check_fails
  Traceback (most recent call last):
    File "/storage/dev/opensourcelib/k2/k2/runner/runner.py", line 304, in run_test_case
      **kwargs)
    File "/storage/dev/opensourcelib/k2/.venv/lib/python3.6/site-packages/zaf/component/factory.py", line 171, in call
      return callable(*args, **kwargs)
    File "/storage/dev/opensourcelib/k2/addons/k2-telnet/telnet/systest/test_telnet.py", line 114, in test_sut_recovery_initiated_when_telnet_connection_check_fails
      queue.get(timeout=10)
    File "/storage/dev/opensourcelib/k2/.venv/lib/python3.6/site-packages/rpyc/core/netref.py", line 200, in __exit__
      return syncreq(self, consts.HANDLE_CTXEXIT, exc)  # can't pass type nor traceback
    File "/storage/dev/opensourcelib/k2/.venv/lib/python3.6/site-packages/rpyc/core/netref.py", line 61, in syncreq
      return conn.sync_request(handler, proxy, *args)
    File "/storage/dev/opensourcelib/k2/.venv/lib/python3.6/site-packages/rpyc/core/protocol.py", line 473, in sync_request
      return self.async_request(handler, *args, timeout=timeout).value
    File "/storage/dev/opensourcelib/k2/.venv/lib/python3.6/site-packages/rpyc/core/async_.py", line 100, in value
      self.wait()
    File "/storage/dev/opensourcelib/k2/.venv/lib/python3.6/site-packages/rpyc/core/async_.py", line 47, in wait
      self._conn.serve(self._ttl)
    File "/storage/dev/opensourcelib/k2/.venv/lib/python3.6/site-packages/rpyc/core/protocol.py", line 386, in serve
      data = self._channel.poll(timeout) and self._channel.recv()
    File "/storage/dev/opensourcelib/k2/.venv/lib/python3.6/site-packages/rpyc/core/channel.py", line 55, in recv
      header = self.stream.read(self.FRAME_HEADER.size)
    File "/storage/dev/opensourcelib/k2/.venv/lib/python3.6/site-packages/rpyc/core/stream.py", line 260, in read
      raise EOFError("connection closed by peer")
  EOFError: connection closed by peer
```

**Comment by per-bohlin** [comment]
@Risca , thank you. Now I understand.

Ideally, I would like to see an addition to the zaf.builtin.blocker, so that that you can force it to block on `AFTER_COMMAND` with command-line arguments similar to how `--blocker-init-enabled` works. Then it could be activated by `--blocker-after-enabled`
and the blocker_helper could be extended with `stop_after_blocking`.

Then the test case could simple call `stop_after_blocking()` after `queue.get()`.

If you don't feel like doing the above changes I would at least change the RemoteBlocker to block on `AFTER_COMMAND` instead of `SUT_RECOVERY_PERFORM`, otherwise you would still run the risk of a race-condition even if it significant reduced.
Blocking on AFTER_COMMAND makes it possible to wait for it after `queue.get()` which eliminates the race condition.
 
Makes sense?

**Comment by Risca** [comment]
> @Risca , thank you. Now I understand.
> 
> Ideally, I would like to see an addition to the zaf.builtin.blocker, so that that you can force it to block on `AFTER_COMMAND` with command-line arguments similar to how `--blocker-init-enabled` works. Then it could be activated by `--blocker-after-enabled`
> and the blocker_helper could be extended with `stop_after_blocking`.
> 
> Then the test case could simple call `stop_after_blocking()` after `queue.get()`.
> 
> If you don't feel like doing the above changes I would at least change the RemoteBlocker to block on `AFTER_COMMAND` instead of `SUT_RECOVERY_PERFORM`, otherwise you would still run the risk of a race-condition even if it significant reduced.
> Blocking on AFTER_COMMAND makes it possible to wait for it after `queue.get()` which eliminates the race condition.
> 
> Makes sense?

Yeap, that makes sense. I'll see if I can fix this up in the coming week :wink:

**Comment by Risca** [comment]
I ran into another issue trying to fix this: sending the unblock command will trigger a disconnect from the client _before_ I get the reply from the unblock command :facepalm: I'll try some more

**Comment by Risca** [comment]
Ok, it was a propblem of which order the callback dispatchers were executed on the remote. The block on exit code didn't work so well if the remote disconnected the client before the client was able to send the STOP command :sweat_smile: Now the init and exit blockers have higher priority than most other dispatchers to ensure they block before any other initialization or shutdown code is run. I believe there is still a slight race condition where the client does not have time to process the RPC call to stop blocking before the client actually disconnects, but it's much harder to reproduce now :/

File: zaf/zaf/builtin/blocker/remote.py
Comment (Risca) [inline zaf/zaf/builtin/blocker/remote.py:14]: Adding `is_async=True` should avoid us getting `EOFException` while handling this request :slightly_smiling_face:

**Comment by Risca** [comment]
With the latest changes I haven't been able to reproduce any server-client race conditions :smiley_cat:

**Comment by Risca** [comment]
> With the latest changes I haven't been able to reproduce any server-client race conditions

I ran this loop for 10-20 minutes:
```bash
while make FEATURE_SYSTESTS=addons/k2-telnet/telnet/systest/test_telnet.py systest_local; do
  echo PASSED;
done
```

**Comment by per-bohlin** [comment]
Looks good. Feel free to rebase and merge.

**Review by per-bohlin — APPROVED** [review]

**Pull request merged**
